src/backup.py

```diff
@@ -128,3 +161,253 @@ def can_use_bucket(self) -> bool:
+
+
+def _get_paths_to_backup(container: ops.Container) -> Iterable[str]:
+    """Get the list of paths that should be in a backup for Synapse.
+
+    Args:
+       container: Synapse Container.
+
+    Returns:
+       Iterable with the list of paths to backup.
+    """
+    paths = []
+    for pattern in BACKUP_FILE_PATTERNS:
+        paths += container.list_files(synapse.SYNAPSE_CONFIG_DIR, pattern=pattern)
+    # Local media if it exists
+    media_dir = os.path.join(synapse.SYNAPSE_DATA_DIR, MEDIA_DIR)
+    if container.exists(media_dir):
+        paths += container.list_files(media_dir, pattern=LOCAL_DIR_PATTERN)
+    return [path.path for path in paths]
+
+
+def _calculate_size(container: ops.Container, paths: Iterable[str]) -> int:
+    """Return the combined size of all the paths given.
+
+    Args:
+        container: Container where to check the size of the paths.
+        paths: Paths to check.
+
+    Returns:
+        Total size in bytes.
+
+    Raises:
+        BackupError: If there was a problem calculating the size.
+    """
+    command = "set -euxo pipefail; du -bsc " + _paths_to_args(paths) + " | tail -n1 | cut -f 1"
+    try:
+        exec_process = container.exec(
+            [BASH_COMMAND, "-c", command],
+            user=synapse.SYNAPSE_USER,
+            group=synapse.SYNAPSE_GROUP,
+        )
+        stdout, stderr = exec_process.wait_output()
+    except (APIError, ExecError) as exc:
+        raise BackupError("Cannot calculate size of paths. du failed.") from exc
+
+    logger.info(
+        "Calculating size of paths. Command: %s. stdout: %s. stderr: %s", command, stdout, stderr
+    )
+
+    try:
+        return int(stdout)
+    except ValueError as exc:
+        raise BackupError("Cannot calculate size of paths. Wrong stdout.") from exc
+
+
+def _build_backup_command(
+    s3_parameters: S3Parameters,
+    backup_id: str,
+    backup_paths: Iterable[str],
+    passphrase_file: str,
+    expected_size: int = int(1e10),
+) -> List[str]:
+    """Build the command to execute the backup.
+
+    Args:
+        s3_parameters: S3 parameters.
+        backup_id: The name of the object to back up.
+        backup_paths: List of paths to back up.
+        passphrase_file: Passphrase to use to encrypt the backup file.
+        expected_size: expected size of the backup, so AWS S3 Client can calculate
+            a reasonable size for the upload parts.
+
+    Returns:
+        The backup command to execute.
+    """
+    bash_strict_command = "set -euxo pipefail; "
+    paths = _paths_to_args(backup_paths)
+    tar_command = f"tar -c {paths}"
+    gpg_command = f"gpg --batch --no-symkey-cache --passphrase-file '{passphrase_file}'"
+    gpg_command += " --symmetric"
+    aws_command = f"{AWS_COMMAND} s3 cp --expected-size={expected_size} - "
+
+    s3_url = _s3_path(
+        prefix=s3_parameters.path, object_name=backup_id, bucket=s3_parameters.bucket
+    )
+    aws_command += f"'{s3_url}'"
```

Comment:
It's minor but consider using the pattern:
```
a = (f"start of the string"
     f"next part of the string"
     )
```
For the long strings. This will also force you to regroup them (`aws_command` is in two places right now)